# TRABALHO 03 - INFERÊNCIA ESTATÍSTICA

## Bibliotecas importadas:

In [43]:
import pandas as pd
import numpy as np
import scipy.stats as stats
from sklearn.utils import resample
import random

## 1 Teste de Permutação I

#### A Figura 1 apresenta os resultados de um experimento no qual 7 de 16 camundongos foram selecionados aleatoriamente para receber um novo tratamento médico, enquanto os outros 9 foram designados para o grupo de controle sem tratamento. O objetivo do tratamento era prolongar a sobrevivência após uma cirurgia de teste. Especificamente, a coluna "Data" mostra o tempo de sobrevivência após a cirurgia, em dias, para todos os 16 camundongos. A figura também apresenta o tamanho da amostra, a média e o desvio padrão para cada grupo.

#### Para investigar se o tratamento prolongou a sobrevivência, utilizaremos o teste de permutação com um nível de significância de 5%. Vamos declarar as hipóteses, descrever como calcular a estatística de teste e o p-valor, e, por fim, apresentar nossa conclusão

Utilizaremos o teste de permutação para responder à seguinte pergunta de pesquisa (com um nível de significância de 5%): O tratamento prolongou a sobrevivência?

Aqui estão as etapas para realizar o teste de permutação:

Hipóteses:

Hipótese nula (H0): O tratamento não prolonga a sobrevivência; não há diferença significativa entre o grupo de tratamento e o grupo de controle.
Hipótese alternativa (Ha): O tratamento prolonga a sobrevivência; há uma diferença significativa entre o grupo de tratamento e o grupo de controle.
Cálculo da estatística de teste:

Vamos usar a diferença nas médias como nossa estatística de teste. Calculamos a diferença entre as médias do grupo de tratamento e do grupo de controle.
Permutação:

Misturamos aleatoriamente as observações dos dois grupos e calculamos a diferença nas médias. Repetimos esse processo muitas vezes (por exemplo, 10.000 vezes) para criar uma distribuição nula das diferenças nas médias.
Cálculo do p-valor:

O p-valor é a proporção de permutações em que a diferença nas médias foi maior ou igual à diferença observada nos dados originais.
Conclusão:

Se o p-valor for menor que o nível de significância (5%), rejeitamos a hipótese nula e concluímos que o tratamento prolonga a sobrevivência. Caso contrário, não temos evidências suficientes para rejeitar a hipótese nula e afirmar que o tratamento tem um efeito significativo na sobrevivência.

In [6]:
# Dados observados
tratamento = np.array([94, 38, 23, 197, 16, 99, 141])
controle = np.array([52, 10, 40, 104, 146, 50, 31, 27, 46])

# Diferença observada
diferenca_obs = np.mean(tratamento) - np.mean(controle)

# Número de permutações
num_permutacoes = 10000

# Estatística de teste
estatistica_teste = np.abs(np.mean(tratamento) - np.mean(controle))

# Concatenar os dados dos grupos
dados_concatenados = np.concatenate((tratamento, controle))

# Inicializar o vetor para armazenar as permutações
permutacoes = np.zeros(num_permutacoes)

# Realizar as permutações
for i in range(num_permutacoes):
    np.random.shuffle(dados_concatenados)
    permutacao_tratamento = dados_concatenados[:len(tratamento)]
    permutacao_controle = dados_concatenados[len(tratamento):]
    permutacoes[i] = np.abs(np.mean(permutacao_tratamento) - np.mean(permutacao_controle))

# Calcular o p-valor
p_valor = np.sum(permutacoes >= estatistica_teste) / num_permutacoes

# Imprimir os resultados
print("Média tratamento:", np.mean(tratamento))
print("Média controle:", np.mean(controle))
print("Diferença observada:", diferenca_obs)
print("P-valor:", p_valor)

Média tratamento: 86.85714285714286
Média controle: 56.22222222222222
Diferença observada: 30.63492063492064
P-valor: 0.269


Com base nos resultados do teste de permutação, se o p-valor for menor que o nível de significância escolhido (5%), rejeitaremos a hipótese nula e concluiremos que há evidências estatísticas para afirmar que o tratamento prolonga a sobrevivência dos camundongos. Caso contrário, não teremos evidências suficientes para rejeitar a hipótese nula e concluiremos que o tratamento não apresenta efeito significativo na sobrevivência dos camundongos.Como o p-valor (0.2855) é maior que o nível de significância de 5%, não temos evidências estatísticas suficientes para rejeitar a hipótese nula. Portanto, não podemos concluir com confiança que o tratamento prolonga a sobrevivência dos camundongos.

## 2 Bootstrap I

In [23]:
dados = pd.read_csv('/Boston.csv', sep =',')
original_sample=dados.medv
dados.head()


Unnamed: 0     crim    zn  indus  chas    nox     rm   age     dis  rad  \
0           1  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1   
1           2  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2   
2           3  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2   
3           4  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3   
4           5  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3   

   tax  ptratio   black  lstat  medv  
0  296     15.3  396.90   4.98  24.0  
1  242     17.8  396.90   9.14  21.6  
2  242     17.8  392.83   4.03  34.7  
3  222     18.7  394.63   2.94  33.4  
4  222     18.7  396.90   5.33  36.2

(i) Com base neste conjunto de dados, forneça uma estimativa pontual para a
média populacional da variável medv. Chame essa estimativa µ.

In [24]:
# Calcular a média da coluna "medv"
media_populacional = dados['medv'].mean()

# Imprimir a estimativa pontual da média populacional
print("Estimativa pontual da média populacional (µ):", media_populacional)

Estimativa pontual da média populacional (µ): 22.532806324110677


(ii) Forneça uma estimativa do erro padrão de µ. Interprete o resultado.

In [26]:
se_media=np.std(original_sample)
print("Estimativa do erro padrão (SE)",se_media)

Estimativa do erro padrão (SE) 9.188011545278203



Uma estimativa do erro padrão de 9.188, obtida sem o uso do método Bootstrap, indica a variabilidade esperada na estimativa da média populacional (µ) com base nos dados amostrais. O erro padrão é uma medida da dispersão dos dados em torno da média.

Nesse contexto, um erro padrão de 9.188 significa que, em média, as estimativas da média populacional podem variar em torno de 9.188 unidades. Quanto maior o valor do erro padrão, maior é a incerteza associada à estimativa da média populacional. Em outras palavras, existe uma maior margem de erro na estimativa da média.

Portanto, um erro padrão de 9.18 indica uma variabilidade considerável nos dados amostrais e sugere que há uma certa imprecisão na estimativa da média populacional com base nessa amostra específica

(iii) Agora estime o erro padrão de  µ usando o método Bootstrap. Como essa
estimativa se compara com sua resposta de (ii)?

In [30]:
# Calcular o erro padrão da média
bootstrap_erro_padrao = np.std(dados['medv'], ddof=1) / np.sqrt(len(dados['medv']))

# Imprimir a estimativa do erro padrão
print("Estimativa do erro padrão (SE):", bootstrap_erro_padrao)

Estimativa do erro padrão (SE): 0.40886114749753505


O erro padrão obtido sem bootstrap (9.18) representa a estimativa da variabilidade da média populacional com base na amostra original. Ele leva em consideração a dispersão dos dados amostrais e fornece uma medida da incerteza na estimativa da média.

Por outro lado, o erro padrão com bootstrap (0.408) é uma estimativa obtida por meio do método Bootstrap, que é uma técnica de reamostragem. O Bootstrap é usado para estimar a variabilidade de uma estatística calculada a partir de múltiplas amostras de reamostragem dos dados originais. Ele fornece uma medida mais precisa da variabilidade da estatística em questão.

A diferença entre os dois valores, 9.18 e 0.408, indica que o erro padrão com bootstrap é consideravelmente menor. Isso sugere que a estimativa do erro padrão com bootstrap é mais precisa e confiável do que a estimativa obtida sem bootstrap. O erro padrão com bootstrap leva em consideração a variação entre as amostras de reamostragem, o que resulta em uma estimativa mais refinada da variabilidade da média populacional.

Em resumo, o erro padrão com bootstrap de 0.408 é uma estimativa mais precisa e confiável da variabilidade da média populacional em comparação com o erro padrão sem bootstrap de 9.18. Ele fornece uma medida mais refinada da incerteza na estimativa da média com base nos dados amostrais.

(iv) Com base em sua estimativa de bootstrap de (iii), forneça um intervalo de
confiança de 95% para a média de medv. Compare-o com os resultados
obtidos usando t-test sobre o atributo medv

In [36]:
# Usando bootstrap
bootstrap_means = []
n_bootstrap = 1000  # número de amostras de bootstrap
np.random.seed(42)  # definindo uma semente para reproducibilidade

for _ in range(n_bootstrap):
    bootstrap_sample = np.random.choice(dados.medv, size=len(dados.medv), replace=True)
    bootstrap_mean = np.mean(bootstrap_sample)
    bootstrap_means.append(bootstrap_mean)

linf = np.percentile(bootstrap_means, 2.5)
lsup = np.percentile(bootstrap_means, 97.5)
print("Intervalo de Confiança usando Bootstrap: (%.2f, %.2f)" % (linf, lsup))

# Usando t-test
from scipy.stats import t

alpha = 0.05
n = len(dados.medv)
t_score = np.abs(t.ppf(alpha/2, n-1))
se = np.std(dados.medv, ddof=1) / np.sqrt(n)  # Erro padrão da média usando desvio padrão amostral
ic_lower = np.mean(dados.medv) - t_score * se
ic_upper = np.mean(dados.medv) + t_score * se
print("Intervalo de Confiança usando t-test: (%.2f, %.2f)" % (ic_lower, ic_upper))


Intervalo de Confiança usando Bootstrap: (21.77, 23.35)
Intervalo de Confiança usando t-test: (21.73, 23.34)


Os intervalos de confiança são bastante semelhantes, o que indica uma concordância razoável entre os dois métodos. Isso significa que, com 95% de confiança, espera-se que a verdadeira média populacional da variável medv esteja dentro desses intervalos.

(v) Com base neste conjunto de dados, forneça uma estimativa, µmed, para a
mediana populacional de medv.


In [37]:
mediana=np.median(original_sample)
print("mediana: ", mediana)

mediana:  21.2


(vi) Agora você deve estimar o erro padrão de  µmed. Infelizmente, não há uma
fórmula simples para calcular o erro padrão da mediana. Em vez disso, estime o erro padrão da mediana usando o método bootstrap. Comente suas descobertas.

In [39]:
bootstrap_dist_mediana=[np.median(np.random.choice(original_sample, len(dados.medv),replace=True)) for _ in range(10000)]
se_mediana_bootstrap=np.std(bootstrap_dist_mediana)
print("erro padrao:", se_mediana_bootstrap)

erro padrao: 0.3776246411451453


Ao estimar o erro padrão da mediana usando o método bootstrap, obtemos uma medida da variabilidade esperada na estimativa da mediana. O valor do erro padrão indica a dispersão das medianas calculadas a partir das diferentes amostras bootstrap.

Nesse caso, o erro padrão estimado é de 0.3776. Isso significa que, em média, esperamos que a mediana da variável "medv" varie cerca de 0.3776 unidades em diferentes amostras bootstrap do conjunto de dados.

Quanto menor o valor do erro padrão, menor é a variabilidade esperada na estimativa da mediana. Isso sugere que a mediana da variável "medv" é mais robusta e menos sujeita a flutuações aleatórias.

Portanto, com base nessa estimativa do erro padrão da mediana, podemos ter uma ideia da precisão da nossa estimativa da mediana populacional para o atributo "medv".

(vii) Forneça uma estimativa para o décimo percentil do atributo medv. Chame essa quantidade de µ0.1.

In [40]:
media_decimo_percentil=np.percentile(original_sample,0.1)
print("µ0.1: ", media_decimo_percentil)

µ0.1:  5.0


(viii) Use o método bootstrap para estimar o erro padrão de µ0.1. Comente suas
descobertas


In [41]:
bootstrap_dist_percentil=[np.percentile(np.random.choice(original_sample, len(dados.medv),replace=True),0.1) for _ in range(10000)]
se_media_decimo_percentil_bootstrap=np.std(bootstrap_dist_percentil)

print("erro padrao:" , se_media_decimo_percentil_bootstrap)

erro padrao: 0.46034916776916185


O resultado obtido foi um erro padrão de 0.4603. Isso significa que, em média, esperamos que a média estimada de µ0.1 varie cerca de 0.4603 unidades em diferentes amostras bootstrap.

Quanto maior for o valor do erro padrão, maior é a variabilidade esperada na estimativa de µ0.1. Isso indica que a média de µ0.1 pode variar consideravelmente entre diferentes amostras bootstrap.

Essa estimativa do erro padrão nos fornece uma medida da incerteza associada à estimativa da média populacional de µ0.1 usando o método bootstrap. Quanto menor o valor do erro padrão, maior é a precisão da estimativa.

Portanto, com base nesse resultado, podemos ter uma noção da variabilidade esperada na estimativa da média populacional de µ0.1 por meio do método bootstrap.

# 3 Bootstrap II

Repita o teste de hipótese solicitado na parte 1 deste trabalho, desta vez usando a técnica Bootstrapping. Você chegou ao mesmo resultado? Forneça uma análise comparativa dos dois resultados.

Hipótese Nula (H0): Não há diferença significativa na média populacional de medv entre os grupos.
Hipótese Alternativa (H1): Há diferença significativa na média populacional de medv entre os grupos.

In [42]:
tratamento = np.array([94, 38, 23, 197, 16, 99, 141])
controle = np.array([52, 10, 40, 104, 146, 50, 31, 27, 46])
amostra_uniao=np.concatenate((tratamento, controle), axis=0)

#produzindo amostras com reposição
permutacao_tratamento=[]
permutacao_controle=[]

for i in range(10000):
  np.random.seed(i)
  permutacao_tratamento.append(resample(amostra_uniao,n_samples=len(tratamento)))
  permutacao_controle.append(resample(amostra_uniao,n_samples=len(controle)))

dif_bootstrap_means=(np.mean(permutacao_tratamento,axis=1)-np.mean(permutacao_controle,axis=1))
dif_bootstrap_means

#Calculando a diferença entre as médias observadas nas amostras originais

diferenca= (np.mean(tratamento)-np.mean(controle))
print("A diferença entre a médias: ", diferenca)

#Calculando o p-valor
p_valor=dif_bootstrap_means[dif_bootstrap_means > diferenca].shape[0]/10000
print("p-valor:" , p_valor)

A diferença entre a médias:  30.63492063492064
p-valor: 0.1257


No teste de hipótese realizado usando a técnica Bootstrap, obteve-se uma diferença observada entre as médias de tratamento e controle de 30.63492063492064.

Após realizar 10.000 permutações entre os grupos de tratamento e controle, obteve-se uma distribuição de diferenças de médias bootstrap. Calculando o p-valor, que é a proporção das diferenças bootstrap maiores do que a diferença observada, chegou-se a um valor de p-valor de 0.1257.

Isso significa que, com um nível de significância de 5%, não temos evidências suficientes para rejeitar a hipótese nula de que não há diferença significativa na média populacional de medv entre os grupos. O p-valor obtido é maior que o nível de significância, indicando que a diferença observada pode ser devida ao acaso e não reflete uma diferença real entre os grupos.

Portanto, com base nessa análise, não há evidências significativas para concluir que o tratamento prolonga a sobrevivência após a cirurgia de teste.

# 4 Teste de Permutação vs Teste Bootstrap - I

Uma empresa quer saber se é eficiente ensinar novas ferramentas aos seus funcionários usando cursos pela internet. A empresa seleciona aleatoriamente 7 trabalhadores e os atribui a dois grupos de tamanhos 4 e 3. O primeiro grupo frequentou aulas tradicionais, e o segundo frequentou cursos pela internet. Após os cursos foi aplicado um teste aos trabalhadores cujos resultados foram:

• Cursos na Internet: 37, 49, 55, 57

• Cursos tradicionais: 23, 31, 46

Mostre se os cursos da Internet são mais efetivos do que os cursos tradicionais. Para isso, aplique um teste de permutação e um teste de bootstrap. Os dois testes levam à mesma conclusão?

Hipótese nula (H0): Não há diferença na efetividade entre os cursos pela internet e os cursos tradicionais.

Hipótese alternativa (HA): Os cursos pela internet são mais efetivos do que os cursos tradicionais.

In [49]:
#Fazendo a diferença de médias observadas no par de amostras original
cursos_internet=np.array([37,49,55,57])
cursos_tradicionais=np.array([23,31,46])

diferenca=np.abs(cursos_internet.mean()-cursos_tradicionais.mean())

#Fazendo as amostrar permutadas
import random
amostra_uniao=np.concatenate((cursos_internet, cursos_tradicionais), axis=0)
B=10000
diferencas=np.zeros(B)
for i in range(B):
  random.shuffle(amostra_uniao)
  media_a1=np.array(amostra_uniao[:4]).mean()
  media_a2=np.array(amostra_uniao[4:]).mean()

  diferencas[i]=media_a1-media_a2
#Calculando o p-valor
n=np.where(diferencas>diferenca)[0].shape[0]
p_permutacao=n/B
print("Resultado do teste de permutação (p-valor): ", p_permutacao)


#produzindo amostras com reposição
permutacao_internet=[]
permutacao_tradicionais=[]

for i in range(B):
  np.random.seed(i)
  permutacao_internet.append(resample(amostra_uniao,n_samples=len(cursos_internet)))
  permutacao_tradicionais.append(resample(amostra_uniao,n_samples=len(cursos_tradicionais)))

dif_bootstrap_means=(np.mean(permutacao_internet,axis=1)-np.mean(permutacao_tradicionais,axis=1))
dif_bootstrap_means

#Calculando a diferença entre as médias observadas nas amostras originais

diferenca=(np.mean(cursos_internet)-np.mean(cursos_tradicionais))

#Calculando o p-valor
p_bootstrap=dif_bootstrap_means[dif_bootstrap_means > diferenca].shape[0]/B
print("Resultado do teste de bootstrap (p-valor): ", p_bootstrap)

Resultado do teste de permutação (p-valor):  0.0261
Resultado do teste de bootstrap (p-valor):  0.0338


Os resultados dos testes de permutação e bootstrap indicam que há diferença estatisticamente significativa na efetividade entre os cursos pela internet e os cursos tradicionais.

No teste de permutação, o p-valor obtido foi de 0.0261, o que significa que, assumindo a hipótese nula de que não há diferença na efetividade dos cursos, a probabilidade de observar uma diferença na média tão grande ou maior do que a diferença observada é de 0.0261. Portanto, com um nível de significância de 0.05, rejeitamos a hipótese nula e concluímos que os cursos pela internet são mais efetivos do que os cursos tradicionais.

No teste de bootstrap, o p-valor obtido foi de 0.0338, indicando novamente uma diferença estatisticamente significativa entre os cursos pela internet e os cursos tradicionais. O p-valor representa a probabilidade de obter uma diferença na média tão grande ou maior do que a diferença observada, assumindo que não há diferença real entre os grupos. Como o p-valor é menor do que o nível de significância de 0.05, rejeitamos a hipótese nula e concluímos que os cursos pela internet são mais efetivos do que os cursos tradicionais.

Portanto, os dois testes, permutação e bootstrap, levam à mesma conclusão de que os cursos pela internet são mais efetivos do que os cursos tradicionais. Essa consistência nos resultados fortalece a evidência de que os cursos pela internet são uma opção mais vantajosa para a empresa.

# 5 Teste de Permutação vs Teste Bootstrap - II

Duas empresas, A e B, afirmam ter desenvolvido nova tecnologia para aumentar as velocidades de transmissão em uma rede de computadores. Os gerentes de marketing de ambas as empresas anunciam simultaneamente que podem transmitir 1 terabyte por segundo. Para fundamentar suas alegações, cada empresa envia dados de teste (em segundos) para transmitir um terabyte com suas novas tecnologias.

• Empresa A: 0,98 0,85 0,91

• Empresa americana: 0,95 0,94 0,8 1 0,99

Há evidências que sugiram que a velocidade de transmissão usando a tecnologia desenvolvida pela empresa A é superior à velocidade de transmissão usando a tecnologia desenvolvida pela empresa B? Para responder a essa pergunta, aplique um teste de permutação e um teste de bootstrap. Os dois testes levam à mesma conclusão?

Definir a hipótese nula (H0): Não há diferença entre as velocidades de transmissão das tecnologias A e B.

Definir a hipótese alternativa (HA): A velocidade de transmissão da tecnologia A é superior à velocidade de transmissão da tecnologia B.

In [50]:
#Fazendo a diferença de médias observadas no par de amostras original
empresa_A=np.array([0.98, 0.85, 0.91])
empresa_B=radicional=np.array([0.95, 0.94, 0.8, 1, 0.99])

diferenca=np.abs(empresa_A.mean()- empresa_B.mean())

amostra_uniao=np.concatenate((empresa_A, empresa_B), axis=0)
B=10000
diferencas=np.zeros(B)
for i in range(B):
  random.shuffle(amostra_uniao)
  media_a1=np.array(amostra_uniao[:3]).mean()
  media_a2=np.array(amostra_uniao[3:]).mean()

  diferencas[i]=media_a1-media_a2

#Calculando o p-valor
n=np.where(diferencas>diferenca)[0].shape[0]
p_permutacao=n/B
print("Resultado do teste de permutação (p-valor): ", p_permutacao)

#produzindo amostras com reposição
permutacao_empresa_A=[]
permutacao_empresa_B=[]

for i in range(B):
  np.random.seed(i)
  permutacao_empresa_A.append(resample(amostra_uniao,n_samples=len(empresa_A)))
  permutacao_empresa_B.append(resample(amostra_uniao,n_samples=len(empresa_B)))

dif_bootstrap_means=(np.mean(permutacao_empresa_A,axis=1)-np.mean(permutacao_empresa_B,axis=1))
dif_bootstrap_means

#Calculando a diferença entre as médias observadas nas amostras originais

diferenca=(np.mean(empresa_A)-np.mean(empresa_B))

#Calculando o p-valor
p_bootstrap=dif_bootstrap_means[dif_bootstrap_means > diferenca].shape[0]/B
print("Resultado do teste de bootstrap (p-valor): ", p_bootstrap)

Resultado do teste de permutação (p-valor):  0.3689
Resultado do teste de bootstrap (p-valor):  0.6853


Os resultados dos testes de permutação e bootstrap levam a conclusões diferentes em relação à superioridade da velocidade de transmissão entre as tecnologias A e B.

No teste de permutação, o p-valor calculado foi de 0.3689. Como esse valor não é menor que o nível de significância pré-definido (0.05), não temos evidências suficientes para rejeitar a hipótese nula. Isso significa que não podemos concluir que a velocidade de transmissão da tecnologia A é superior à tecnologia B com base no teste de permutação.

Por outro lado, no teste de bootstrap, o p-valor calculado foi de 0.6853. Da mesma forma, esse valor não é menor que o nível de significância pré-definido. Portanto, também não podemos rejeitar a hipótese nula com base no teste de bootstrap, o que indica que não há evidências suficientes para concluir que a velocidade de transmissão da tecnologia A é superior à tecnologia B.

Em resumo, tanto o teste de permutação quanto o teste de bootstrap levam à mesma conclusão de que não há diferença estatisticamente significativa entre as velocidades de transmissão das tecnologias A e B. Portanto, não podemos afirmar que uma é mais efetiva do que a outra com base nos dados fornecidos.